### ISSUE TITLE
Shared repository schedule possibility

### ISSUE THREAD

**Issue by flpelluz** [renovatebot/config-help#841]
### Which Renovate are you using?

Renovate App

### Which platform are you using?

GitHub Enterprise

### Have you checked the logs? Don't forget to include them if relevant

Yes, I checked the docs https://docs.renovatebot.com/self-hosted-configuration/ but wasn't able to find an example. 

### What would you like to do?

Hello! 

So my goal is to have a centralised settings for a specific list of repositories. I got the centralised part working, where each repo is extending some base-config.json, but I would also like to define a specific schedule per repository in the base-config.json. 

I know I could specify the schedule in each repository own renovate.json file, I need this file to extend the base-config anyway. However,  this prevents from getting a quick view of the different schedules set for each repo. Hence, having this schedule in the base-config.json instead would be great. 

Is this possible? 

Thanks!

**Comment by rarkins** [comment]
It's not possible, and doesn't really fit with the concept of extending presets because the preset would need to have awareness of the repositories that extend it. However you could define per-repo settings in your bot's config.js and get some visibility that way.

**Comment by flpelluz** [comment]
Thank you for the quick answer @rarkins. 

> However you could define per-repo settings in your bot's config.js and get some visibility that way.

I don't think I follow how this differ from the current set up I do have. Could you give me an example?

**Comment by flpelluz** [comment]
Besides, I would like to apply a schedule to all the packageRules. For instance, the second week of the month. Then, during that week, each rule would have its own schedule, like on Monday. Is this possible or would I need to specify the schedule to be instead : "on Monday of the second week of the month".

**Comment by rarkins** [comment]
Presets are for user/repo configs. config.js is the bot admin settings.

Package rules can define schedules but not vice versa.

I think it's likely you have one or two mistaken assumptions about Renovate which is making this complex. Can you back up a step or two and describe in plain English what you want to achieve? Ie forget implementation details about package rules, presets and schedules and just tell me your business/workflow requirements. 

Remember you have a separation of concerns between what the bot admin can control vs what admins of repos can override. Typical repo admins have complete control over renovate functionality such as schedules and rules.

**Comment by flpelluz** [comment]
Yes, you were correct. I was actually missing some context which I just got 😅 . 

So, in plain words, we have the renovate Github app set up for all our repos. Then, each repo has its own renovate.json file. This is the basic set-up. We don't have a specific control set up for the bot admin in any other way. 

We are in the process of sharing configs as much as possible. For that, we have created a repo containing different config files. Each config file might make use of one or many presets plus other settings. The, idea is that repositories can extend the config file they need from this repo instead. This is working already fine. 

Then, in my team, we are responsible for a bunch of repos within the org. These repos have the same settings. So, I have created a config file in the shared repo so that my team repos will extend this config file instead. This config file has package rules, where each rule has its own schedule, for instance, on Mondays. This is also working fine. 

Now comes what I want to achieve. On top of this, I would like renovate to run once a month, on different weeks, for each repo we are responsible for. Then, every time it runs, it would extend the config file which contains different package rules which are supposed to run on different days of that week.

Hopefully it is clearer now.

**Comment by viceice** [comment]
This is not possible with the renovate app, as you can only have one config per repo, which can extend other configs.

Package rules are all merged together if they apply to a package, so last schedule wins. As schedule is an array, you can defined multiple schedules in one package rule. eg:

```js
...
packageRules: [
...
  {
    packageNames: ['renovate'],
    extends: ["github>org/repo:preset"],
    schedule: ["on mondays", "on fridays"] 
  }
...
],
...
```

**Comment by flpelluz** [comment]
@viceice thanks for the reply. So, how does schedule behave if the config file of a repo was this one below?   

```json
{
"extends": ["github>our_org/renovate-config:our_team_config"],
"schedule": [every 3 months on the first day of the month]
}
```

Then in the shared repo, the team_config file would be: 

```json
...
packageRules: [
...
  {
    packageNames: ['renovate'],
    schedule: ["on mondays"]
  }
...
], 
...
```

**Comment by flpelluz** [comment]
The solution above would suffice for us. But now that it's clear, what I'm after would be this: 

repoA renovate.json

```js
{
"extends": ["github>our_org/renovate-config:our_team_config"]
}
```
Then in the shared repo, our_team_config file would be:

```js
repositories: [
{ 
"repositoryName": "repoA", 
schedule: "every 3 months on the first day of the month"
}, 
{ 
"repositoryName": "repoB", 
schedule: "every 2 months on the fifth day of the month"
}],
packageRules: [
...
  {
    "packageNames": ["renovate"],
    "schedule": ["on mondays"]
  }
...
], 
...
```

**Comment by viceice** [comment]
the package rule schedule will win over the global schedule.

the above shared config will not work, you can simply add more shared config files, or multiple presets in one file.
the `repositories` kes is an admin only config, so not usable with the renovate app.

so every repo should extend one shared config and one repo specific preset.

**shared**
```js
{
  shared: {
  ...
  },
  repoA: {
  ...
  },
  repoB: {
  ...
  }
}
```

**repoA config**
```js
{
  extends: ['github>org/repo:shared', 'github>org/repo:repoA ],
  ...
}
```

you can also use multiple files:
**repoA config**
```js
{
  extends: ['github>org/repo/shared', 'github>org/repo/repoA'],
  ...
}
```

this requires `default.json` and `repoA.json` in you shared repo.

**Comment by viceice** [comment]
You can make it even shorter if all configs are on same platform

repoA config
```js
{
  extends: ['org/repo/shared', org/repo/repoA'],
  ...
}
```

this requires `default.json` and `repoA.json` in your shared `org/repo`.

**Comment by flpelluz** [comment]
@viceice thanks for the reply! Great to know. This definitely increases the visibility and the maintainability because everything is just in one file. Question, because all the repos in the file use the same packageRules, could I even go one step further and do this? 

```js
{
  shared: {
    packageRules: [
    ...
    {
      "packageNames": ["renovate"],
      "schedule": ["on mondays"]
    }
    ...
    ], 
  },
  repoA: {
  extends: ['org/repo/shared'] 
  },
  repoB: {
  extends: ['org/repo/shared'] 
  }
}
```

then, repo A config would only be: 

{
  extends: ['org/repo/repoA']
}

**Comment by viceice** [comment]
should work

**Comment by flpelluz** [comment]
Great to know, thanks for the reply! 

> the package rule schedule will win over the global schedule.

Then, finally, if the package rule schedule will win over the global schedule, is there any way I could have different schedules for when renovate will check a repo but the same weekly schedule for the packages?

**Comment by flpelluz** [comment]
> the package rule schedule will win over the global schedule.

Then, finally, if the package rule schedule will win over the global schedule, is there any way I could have different schedules for when renovate will check a repo but the same weekly schedule for the packages?

**Comment by viceice** [comment]
but if you use a single file, you need to use
```js
{
  extends: ['org/repo:repoA']
}
```

and have a `default.json` in `org/repo`.

**Comment by viceice** [comment]
```js
{
  shared: {
    packageRules: [
    ...
    {
      "packageNames": ["renovate"],
      "schedule": ["on mondays"]
    }
    ...
    ], 
  },
  repoA: {
    extends: ['org/repo:shared'] 
  },
  repoB: {
    extends: ['org/repo:shared'] 
  }
}
```

**Comment by viceice** [comment]
> > the package rule schedule will win over the global schedule.
> 
> Then, finally, if the package rule schedule will win over the global schedule, is there any way I could have different schedules for when renovate will check a repo but the same weekly schedule for the packages?

then you need to override the packageRule with new schedule too.

**Comment by flpelluz** [comment]
Ok, I understand. It's a pity then because the only difference between each repo is the week of the month in the schedule of the package. But the packageRules are actually the same. This means losing the advantage of using the shared settings for the packageRules of above. 

{
 shared: {
    ... 
 },
 repoA: {
   extends: ['org/repo:shared'],
   packageRules: [
    ...
    {
      "packageNames": ["renovate"],
      "schedule": ["every FIRST week of the month on Monday"]
    },
    {
      packageNames: ["Typescript"],
      schedule: ["every FIRST week of the month on Tuesday"]
    }, 
    ...
    ], 
  },
  repoB: {
   extends: ['org/repo:shared'],
   packageRules: [
    ...
    {
      packageNames: ["Renovate"],
      schedule: ["every SECOND week of the month on Monday"]
    }, 
    {
      packageNames: ["Typescript"],
      schedule: ["every SECOND week of the month on Tuesday"]
    }, 
    ...
    ]
  }
}

**Comment by viceice** [comment]
That's the way it works today.

**Comment by flpelluz** [comment]
I see. I guess a good compromise would be then: 

```js 
{
 shared: {
    ... 
    packageRules: [
    {
      packageNames: ["renovate"]
    },
    {
      packageNames: ["Typescript"]
    }, 
  ...
 },
 repoA: {
   extends: ['org/repo:shared'],
   schedule: ["every SECOND week of the month on Monday"]
  },
  repoB: {
   extends: ['org/repo:shared'],
   schedule: ["every SECOND week of the month on Monday"]
  }
...
}
```

**Comment by flpelluz** [comment]
Thanks a lot for confirming and answering all my questions. If you guys ever decide to implement it, let me know :)

**Comment by flpelluz** [comment]
Hello @viceice, sorry to re-open this, but the config doesn't seem to be working. 

Instead of using the default.json, I've created a separate file called: `team.json` in the shared repo. Then, I've assumed it should work in the same way, but there is something going on.  

team.json file in the shared repo. 

```
{
 shared: {
   packageRules: [
    {
      packageNames: ["renovate"]
    },
    {
      packageNames: ["Typescript"]
    }, 
    ...
  ]
 },
 repoA: {
   extends: ["github>org/renovate-config/team:shared"],
   schedule: ["every FIRST week of the month on Monday"]
  },
 repoB: {
   extends: ["github>org/renovate-config/team:shared"],
   schedule: ["every SECOND week of the month on Monday"]
  }
 ...
}
```


repoA renovate.json file, I have used: 
```js
{
 "extends": ["github>smartlyio/renovate-config/team:repoA"]
}
```

Thanks for looking into it! 

Fran

**Comment by viceice** [comment]
@flpelluz Do you see any warnings or errors in renovate logs?

**Comment by flpelluz** [comment]
@viceice yep, I got an issue opened by Renovate in the repoA saying:  `Cannot find preset's package (github>smartlyio/renovate-config/team:repoA)`. We are using the Github renovate app, so I'm not sure if we have access to other types of logs.

**Comment by viceice** [comment]
@flpelluz you can use https://app.renovatebot.com/dashboard, there should be the full debug log

**Comment by flpelluz** [comment]
Ok, thanks. I wasn't aware of this dashboard. I will give it a look and will come back to you if I cannot figure it out.

**Comment by flpelluz** [comment]
So, based on the logs, it seemed that it failed, but there has been an internal migration which has fixed the error by itself? 
```
INFO: Repository started
{
  "renovateVersion": "23.18.2"
}
DEBUG: Using localDir: /mnt/renovate/gh/org/repoA
DEBUG: Repository cache is valid
DEBUG: initRepo("org/repoA")
DEBUG: No dangling containers to remove
DEBUG: org/repoA default branch = master
DEBUG: Using personal access token for git init
DEBUG: resetMemCache()
DEBUG: checkOnboarding()
DEBUG: isOnboarded()
DEBUG: findFile(renovate.json)
DEBUG: Initializing git repository into /mnt/renovate/gh/org/repoA
DEBUG: git clone completed
{
  "durationMs": 976
}
DEBUG: latest commit
{
  "latestCommitDate": "2020-09-07T11:51:55+03:00"
}
DEBUG: Setting git author name
{
  "gitAuthorName": "Renovate Bot"
}
DEBUG: Setting git author email
{
  "gitAuthorEmail": "bot@renovateapp.com"
}
DEBUG: config file exists
DEBUG: ensureIssueClosing(Action required: Add a Renovate config)
DEBUG: Retrieving issueList
DEBUG: Retrieved 1 issues
DEBUG: Repo is onboarded
DEBUG: Found renovate.json config file
DEBUG: Repository config
{
  "fileName": "renovate.json",
  "config": {
    "extends": [
      "github>org/renovate-config/team:repoA"
    ]
  }
}
DEBUG: migrateAndValidate()
DEBUG: No config migration necessary
DEBUG: massaged config
{
  "config": {
    "extends": [
      "github>org/renovate-config/team:repoA"
    ]
  }
}
DEBUG: migrated config
{
  "config": {
    "extends": [
      "github>org/renovate-config/team:repoA"
    ]
  }
}
DEBUG: GitHub 404
{
  "url": "https://api.github.com/repos/org/renovate-config/team/contents/repoA.json"
}
DEBUG: Failed to retrieve repoA.json from repo
{
  "statusCode": 404
}
DEBUG: Preset fetch error
{
  "preset": "github>org/renovate-config/team:repoA",
  "err": {
    "message": "dep not found",
    "stack": "Error: dep not found\n    at fetchJSONFile (/home/ubuntu/renovateapp/node_modules/renovate/dist/config/presets/github/index.js:22:15)\n    at runMicrotasks ()\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)"
  }
}
INFO: Throwing preset error
{
  "validationError": "Cannot find preset's package (github>org/renovate-config/team:repoA)"
}
INFO: Repository has invalid config
{
  "error": {
    "validationError": "Cannot find preset's package (github>org/renovate-config/team:repoA)"
  }
}
DEBUG: raiseConfigWarningIssue()
DEBUG: getBranchPr(renovate/configure)
DEBUG: findPr(renovate/configure, undefined, open)
DEBUG: Retrieving PR list
DEBUG: Retrieved 198 Pull Requests
DEBUG: ensureIssue(Action Required: Fix Renovate Configuration)
DEBUG: Issue is open and up to date - nothing to do
DEBUG: Repository timing splits (milliseconds)
{
  "splits": {},
  "total": 6862
}
DEBUG: http statistics
{
  "hostStats": [
    "api.github.com, 5 requests, 919ms average"
  ],
  "totalRequests": 5
}
INFO: Repository finished
{
  "durationMs": 6862
} 
```

**Comment by flpelluz** [comment]
Should I close the failing issue opened by Renovate in the repo and kick in the bootstrap config in the repo somehow so that it picks up the new settings?

**Comment by viceice** [comment]
I think we have a bug `https://api.github.com/repos/org/renovate-config/team/contents/repoA.json`

i think url must be `https://api.github.com/repos/org/renovate-config/contents/team.json` and `repoA` should be a key in `team.json`

I'll try to debug.

/cc @rarkins expected behavior?

**Comment by flpelluz** [comment]
Thanks @viceice I'll stay tuned for your reply. Yes, I have also noticed about the URL that should be used instead. `repoA` is already a key in team.json. So, that should be alright. I can definitely change the URL in case addressing the bug becomes a bigger task.

**Comment by rarkins** [comment]
Seems like a bug to me too

**Comment by flpelluz** [comment]
@viceice @rarkins thank you for the confirmation. Do you have any estimates on the fix? No hurry on my side at all, just wondering if I should change to use the wrong url or wait for a fix in case it's a matter of a couple of days.

**Comment by flpelluz** [comment]
Also, when we are on it, I got a related question to the main topic of this issue. 

If my config file looks like the one below, do the top `extends, `lockFileMaintenance` and `timezone` apply to all presets or do I need to include them inside each preset specifically? As you can see, there is no extends for the `repoC` assuming that it would inherit the settings defined at the top. 

```js
{
 extends: [config:base],
 "lockFileMaintenance": {
      "enabled": true
    },
 "timezone": "Europe/Berlin",
 shared: {
   packageRules: [
    {
      packageNames: ["renovate"]
    },
    {
      packageNames: ["Typescript"]
    }, 
    ...
  ]
 },
 repoA: {
   extends: ["github>org/renovate-config/team:shared"],
   schedule: ["every 2nd week on Monday"]
  },
 repoB: {
   extends: ["github>org/renovate-config/team:shared"],
   schedule: ["every 3rd week on Monday"]
  }
 repoC: {
  ...
 }
}
```

Thanks!

**Comment by viceice** [comment]
@flpelluz As i can see in our code, it looks like this is currently by design and not easily changeable.

Can you try this:

*org/renovate-config/teams.json*
```js
{
 shared: {
   packageRules: [
    {
      packageNames: ["renovate"]
    },
    {
      packageNames: ["Typescript"]
    }, 
    ...
  ]
 },
 repoA: {
   extends: ["github>org/renovate-config:team/shared"],
   schedule: ["every 2nd week on Monday"]
  },
 repoB: {
   extends: ["github>org/renovate-config:team/shared"],
   schedule: ["every 3rd week on Monday"]
  }
 ...
}
```

*org/teamA/renovate.json*
```js
{
 "extends": ["github>org/renovate-config:team/repoA"]
}
```

**Comment by flpelluz** [comment]
@viceice sure, I will give it a try right now and let you know if it works. Thanks!

**Comment by viceice** [comment]
> Also, when we are on it, I got a related question to the main topic of this issue.
> 
> If my config file looks like the one below, do the top `extends`, `lockFileMaintenance` and `timezone` apply to all presets or do I need to include them inside each preset specifically? As you can see, there is no extends for the `repoC` assuming that it would inherit the settings defined at the top.
> 
> ```js
> {
>  extends: [config:base],
>  lockFileMaintenance: {
>       "enabled": true
>  },
>  timezone: "Europe/Berlin",
>  shared: {
>    packageRules: [
>     {
>       packageNames: ["renovate"]
>     },
>     {
>       packageNames: ["Typescript"]
>     }, 
>     ...
>   ]
>  },
>  repoA: {
>    extends: ["github>org/renovate-config/team:shared"],
>    schedule: ["every 2nd week on Monday"]
>   },
>  repoB: {
>    extends: ["github>org/renovate-config/team:shared"],
>    schedule: ["every 3rd week on Monday"]
>   }
>  repoC: {
>   ...
>  }
> }
> ```
> 
> Thanks!

The toplevel settings are ignored, because we explicitly use only the sub preset objects

**Comment by flpelluz** [comment]
@viceice roger, thanks for the confirmation on the related question above.  

By the way, the job hasn't run for one hour, so I'm not sure if the new change works. I will let you know as soon as it shows in the dashboard. Is there a way to trigger it somehow?

**Comment by flpelluz** [comment]
@viceice the job just ran and it worked. No errors. Great, thank you for all the support and help! By the way, is there any way to re-trigger the jobs from another hidden dashboard 😅 ? Also, how often do they run? I thought it would be on the hour but it doesn't seem anymore the case :)

**Comment by viceice** [comment]
Renovate App normally runs hourly, but also runs on changes to the repo.

You can use the `dependencyDashboard` to trigger a run (there is a checkbox), but better you poush a change to default branch to trigger a run, otherwise caching can skip you repo.

**Comment by flpelluz** [comment]
@viceice understood and good to know. Still, how do I access the dependencyDashboard?

**Comment by viceice** [comment]
Enable it in your shared config, then renovate schould open an issue named dependency dashboard on next run.

**Comment by flpelluz** [comment]
Ah ok, great, I will do it. Thanks for the reply and for addressing all the questions!
